presto-main/src/main/java/com/facebook/presto/operator/WorkProcessor.java

```diff
@@ -127,6 +127,33 @@
 
     interface Transformation<T, R>
     {
+        /**
+         * Processes input elements and returns current transformation state.
+         *
+         * @param elementOptional an element to be transformed. Will be empty
+         * when there are no more elements. In such case transformation should
+         * finish processing and flush any remaining data.
+         * @return a current transformation state.
+         * <ul>
+         * <li>if state is {@link ProcessorState#finished()} then transformation has finished
+         * and {@link Transformation#process(java.util.Optional)} won't be called again;</li>
+         * <li>if state is {@link ProcessorState#needsMoreData} then transformation requires
+         * more data in order to continue. {@link Transformation#process(java.util.Optional)}
+         * will be called with a new input element or with {@link Optional#empty()} if there
+         * are no more elements;</li>
+         * <li>if state is {@link ProcessorState#yield()} then transformation has yielded.
+         * {@link Transformation#process(java.util.Optional)} will be called again with
+         * the same input element;</li>
+         * <li>if state is {@link ProcessorState#blocked} then transformation is blocked.
+         * {@link Transformation#process(java.util.Optional)} will be called again with
+         * the same input element after blocked future is done;</li>
+         * <li>if state is {@link ProcessorState#ofResult(Object, boolean)} then transformation
+         * has produces a result. {@link Transformation#process(java.util.Optional)} will be called
+         * again with a new element (or with {@link Optional#empty()} if there are no more elements)
+         * if <code>needsMoreData</code> {@link ProcessorState#ofResult(Object, boolean)} parameter
+         * is <code>true</code>.</li>
```

Comment:
When you read this, you cannot be sure that you parsed this sentence the way it was intended to.. Split this into 2 list items (or sub-items), depending on the `needsMoreData` value.

Also, `com.facebook.presto.operator.WorkProcessor.ProcessorState#ofResult(T, boolean)` doesn't actually represent a processing state. It represents a transition (outputting value) to a state. `ProcessorState` should be renamed to `ProcessorResult` and `ProcessorState#Type` should be renamed to `ProcessorState`.
(might be as a follow-up)